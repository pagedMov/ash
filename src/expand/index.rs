use crate::{prelude::*, shellenv::LashVal};

pub fn expand_index(pair: Pair<Rule>,lash: &mut Lash) -> LashResult<String> {
	let mut inner = pair.step(1).unpack()?.into_inner().peekable();
	let arr_name = inner.next().unpack()?;

	let array = lash.borrow_vars().get_var(arr_name.as_str());
	let mut cur_val = match array {
		Some(LashVal::Array(vec)) => Some(LashVal::Array(vec)),
		_ => return Ok(String::new()), // If not an array, return empty
	};

	while let Some(index) = inner.next() {
		let idx = index.as_str().parse::<usize>().map_err(|_| Low(LashErrLow::IndexErr(format!("Index '{}' out of range for array '{}'",index,arr_name.as_str()))))?;
		cur_val = match cur_val {
			Some(LashVal::Array(vec)) => vec.get(idx).cloned(),
			_ => return Ok(String::new()), // Invalid nesting (e.g., indexing a non-array)
		};
	}

	Ok(cur_val.map_or_else(String::new, |val| val.to_string()))
}
