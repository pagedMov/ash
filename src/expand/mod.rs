pub mod brace;
pub mod cmdsub;
pub mod dispatch;
pub mod glob;
pub mod index;
pub mod misc;
pub mod string;
