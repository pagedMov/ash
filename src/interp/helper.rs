use crate::{event::ShellError, interp::{parse::Span, token::{Tk, WdFlags, WordDesc, BUILTINS, CMDSEP, KEYWORDS, REGEX, WHITESPACE}}};
use libc::STDERR_FILENO;
use log::{debug,trace};
use nix::unistd::dup2;
use std::{collections::VecDeque, fs, io, os::fd::{AsRawFd, BorrowedFd}};

pub trait StrExtension {
	fn has_unescaped(&self, pat: &str) -> bool;
}

impl StrExtension for str {
	/// Checks to see if a string slice contains a specified unescaped character. This method assumes that '\\' is the escape character.
	///
	fn has_unescaped(&self, pat: &str) -> bool {
		let mut chars = self.chars().collect::<VecDeque<char>>();
		let mut working_pat = String::new();
		let mut escaped = false;

		while let Some(ch) = chars.pop_front() {
			if !escaped && working_pat == pat {
				return true;
			}
			match ch {
				' ' | '\t' => {
					// Check for unescaped match when encountering a space/tab
					// Reset for next segment
					escaped = false;
					working_pat.clear();
				}
				'\\' => {
					escaped = true;
				}
				_ => {
					working_pat.push(ch);
				}
			}
		}

		// Check for unescaped match at the end of the string
		!escaped && working_pat == pat
	}
}

pub fn get_stderr<'a>() -> BorrowedFd<'a> {
	unsafe { BorrowedFd::borrow_raw(STDERR_FILENO) }
}

// This is used when pesky system calls want to emit their own errors
// Which ends up being redundant, since rsh has it's own error reporting system
// Kind of hacky but fuck it
// It works by taking the function as an argument and then executing it in
// a context where stderr is redirected to /dev/null.
pub fn suppress_err<F: FnOnce() -> T, T>(f: F) -> T {
	let stderr = io::stderr();
	let stderr_fd = stderr.as_raw_fd();
	let devnull = fs::OpenOptions::new().write(true).open("/dev/null").unwrap();
	let devnull_fd = devnull.as_raw_fd();

	dup2(devnull_fd, stderr_fd).unwrap();
	let result = f();
	dup2(stderr_fd,stderr_fd).unwrap();
	result
}

pub fn get_stdout<'a>() -> BorrowedFd<'a> {
	unsafe { BorrowedFd::borrow_raw(STDERR_FILENO) }
}

pub fn get_delimiter(wd: &WordDesc) -> char {
	let flags = wd.flags;
	match () {
		_ if flags.contains(WdFlags::IN_BRACE) => '}',
		_ if flags.contains(WdFlags::IN_PAREN) => ')',
		_ => unreachable!("No active delimiter found in WordDesc flags"),
	}
}
pub fn is_brace_expansion(text: &str) -> bool {
	REGEX["brace_expansion"].is_match(text) &&
		REGEX["brace_expansion"].captures(text).unwrap()[1].is_empty()
}
pub fn delimited(wd: &WordDesc) -> bool {
	wd.flags.contains(WdFlags::IN_PAREN)
}
pub fn cmdsep(c: &char) -> bool {
	CMDSEP.contains(c)
}
pub fn keywd(wd: &WordDesc) -> bool {
	KEYWORDS.contains(&wd.text.as_str()) && !wd.flags.contains(WdFlags::IS_ARG)
}
pub fn builtin(wd: &WordDesc) -> bool {
	BUILTINS.contains(&wd.text.as_str()) && !wd.flags.contains(WdFlags::IS_ARG)
}
pub fn wspace(c: &char) -> bool {
	WHITESPACE.contains(c)
}
pub fn quoted(wd: &WordDesc) -> bool {
	wd.flags.contains(WdFlags::SNG_QUOTED) || wd.flags.contains(WdFlags::DUB_QUOTED)
}
pub fn check_redirection(c: &char, chars: &mut VecDeque<char>) -> bool {
	chars.push_front(*c);
	let mut test_chars = chars.clone();
	let mut test_string = String::new();

	while let Some(c) = test_chars.pop_front() {
		if c.is_whitespace() || !matches!(c, '&' | '0'..='9' | '>' | '<') {
			break;
		}
		test_string.push(c);
	}

	if REGEX["redirection"].is_match(&test_string) {
		true
	} else {
		chars.pop_front();
		false
	}
}

pub fn process_redirection(
	word_desc: &mut WordDesc,
	chars: &mut VecDeque<char>,
) -> Result<WordDesc, ShellError> {
	let mut redirection_text = String::new();
	while let Some(c) = chars.pop_front() {
		debug!("found this char in redirection: {}",c);
		if !matches!(c, '&' | '0'..='9' | '>' | '<') {
			chars.push_front(c);
			break;
		}
		redirection_text.push(c);
	}

	debug!("returning this word_desc text: {}",redirection_text);
	Ok(WordDesc {
		text: redirection_text,
		span: word_desc.span,
		flags: WdFlags::IS_OP,
	})
}
pub fn finalize_delimiter(word_desc: &WordDesc) -> Result<WordDesc, ShellError> {
	let mut updated_word_desc = word_desc.clone();

	if word_desc.contains_flag(WdFlags::IN_BRACE) {
		updated_word_desc = updated_word_desc.remove_flag(WdFlags::IN_BRACE);
	} else if word_desc.contains_flag(WdFlags::IN_PAREN) {
		updated_word_desc = updated_word_desc.remove_flag(WdFlags::IN_PAREN);
	}

	Ok(updated_word_desc)
}

pub fn count_spaces(chars: &mut VecDeque<char>) -> usize {
	let mut count = 1;
	let mut buffer = VecDeque::new();
	while let Some(ch) = chars.pop_front() {
		if ch.is_whitespace() {
			buffer.push_back(ch);
			count += 1;
		} else {
			chars.push_front(ch);
			while let Some(ch) = buffer.pop_back() {
				chars.push_front(ch);
			}
			break
		}
	}
	count
}

pub fn finalize_word(word_desc: &WordDesc, tokens: &mut VecDeque<Tk>) -> Result<WordDesc,ShellError> {
	let mut word_desc = word_desc.clone();
	let span = Span::from(word_desc.span.end,word_desc.span.end);
	trace!("finalizing word `{}` with flags `{:?}`",word_desc.text,word_desc.flags);
	if !word_desc.text.is_empty() {
		if keywd(&word_desc) {
			word_desc = word_desc.add_flag(WdFlags::KEYWORD);
		} else if builtin(&word_desc) {
			word_desc = word_desc.add_flag(WdFlags::BUILTIN);
		}
		if word_desc.flags.contains(WdFlags::EXPECT_IN) && matches!(word_desc.text.as_str(), "in") {
			debug!("setting in flag to keyword");
			word_desc = word_desc.remove_flag(WdFlags::IS_ARG);
			word_desc = word_desc.add_flag(WdFlags::KEYWORD);
		}
		// TODO: this logic is really flimsy, probably needs to be refactored
		tokens.push_back(Tk::from(word_desc)?);
	}

	// Always return a fresh WordDesc with reset state
	Ok(WordDesc {
		text: String::new(),
		span,
		flags: WdFlags::empty(),
	})
}
