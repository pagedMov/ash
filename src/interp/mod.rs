pub mod parse;
pub mod helper;
pub mod token;
pub mod expand;
pub mod debug;
