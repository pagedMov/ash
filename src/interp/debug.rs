use std::fmt;

use crate::interp;
use interp::parse::{NdType,Node};

impl fmt::Display for Node {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fn write_tree(node: &Node, f: &mut fmt::Formatter<'_>, indent: usize) -> fmt::Result {
            let prefix = "|   ".repeat(indent);
            writeln!(f, "{}Node (span: {:?})", prefix, node.span)?;

            match &node.nd_type {
                NdType::Root { deck } => {
                    writeln!(f, "{}Root", prefix)?;
                    for node in deck {
                        write_tree(node, f, indent + 1)?;
                    }
                }
                NdType::If { cond_blocks, else_block } => {
                    writeln!(f, "{}If", prefix)?;
                    for block in cond_blocks {
                        writeln!(f, "{}|--- Condition", prefix)?;
                        write_tree(&block.condition, f, indent + 2)?;
                        writeln!(f, "{}|--- Body", prefix)?;
                        write_tree(&block.body, f, indent + 2)?;
                    }
                    if let Some(else_block) = else_block {
                        writeln!(f, "{}|--- Else", prefix)?;
                        write_tree(else_block, f, indent + 1)?;
                    }
                }
                NdType::For { loop_vars, loop_arr, loop_body } => {
                    let var_texts: Vec<_> = loop_vars.iter().map(|var| var.text()).collect();
                    let arr_texts: Vec<_> = loop_arr.iter().map(|arr| arr.text()).collect();
                    writeln!(f, "{}For", prefix)?;
                    writeln!(f, "{}|--- Loop Vars: {:?}", prefix, var_texts)?;
                    writeln!(f, "{}|--- Loop Array: {:?}", prefix, arr_texts)?;
                    writeln!(f, "{}|--- Body", prefix)?;
                    write_tree(loop_body, f, indent + 1)?;
                }
                NdType::Loop { condition, logic } => {
                    writeln!(f, "{}Loop (condition: {})", prefix, condition)?;
                    writeln!(f, "{}|--- Condition", prefix)?;
                    write_tree(&logic.condition, f, indent + 1)?;
                    writeln!(f, "{}|--- Body", prefix)?;
                    write_tree(&logic.body, f, indent + 1)?;
                }
                NdType::Case { input_var, cases } => {
                    writeln!(f, "{}Case (input: {:?})", prefix, input_var.text())?;
                    for case in cases {
                        writeln!(f, "{}|--- Pattern: {:?}", prefix, case.0)?;
                        write_tree(case.1, f, indent + 1)?;
                    }
                }
                NdType::Select { select_var, opts, body } => {
                    let opts_texts: Vec<_> = opts.iter().map(|opt| opt.text()).collect();
                    writeln!(
                        f,
                        "{}Select (input: {:?}, options: {:?})",
                        prefix, select_var.text(), opts_texts
                    )?;
                    write_tree(body, f, indent + 1)?;
                }
                NdType::Pipeline { left, right, both: _ } => {
                    writeln!(f, "{}Pipeline", prefix)?;
                    writeln!(f, "{}|--- Left", prefix)?;
                    write_tree(left, f, indent + 1)?;
                    writeln!(f, "{}|--- Right", prefix)?;
                    write_tree(right, f, indent + 1)?;
                }
                NdType::Chain { left, right, op } => {
                    writeln!(f, "{}Chain (operator span: {:?})", prefix, op.span)?;
                    writeln!(f, "{}|--- Left", prefix)?;
                    write_tree(left, f, indent + 1)?;
                    writeln!(f, "{}|--- Right", prefix)?;
                    write_tree(right, f, indent + 1)?;
                }
                NdType::BraceGroup { body } => {
                    writeln!(f, "{}BraceGroup", prefix)?;
                    write_tree(body, f, indent + 1)?;
                }
                NdType::Subshell { body } => {
                    writeln!(f, "{}Subshell (body: {:?})", prefix, body)?;
                }
                NdType::FuncDef { name, body } => {
                    writeln!(f, "{}Function Definition (name: {})", prefix, name)?;
                    write_tree(body, f, indent + 1)?;
                }
                NdType::Assignment { name, value } => {
                    writeln!(f, "{}Assignment (name: {}, value: {:?})", prefix, name, value)?;
                }
                NdType::Command { argv } | NdType::Builtin { argv } => {
                    let argv_texts: Vec<_> = argv.iter().map(|arg| arg.text()).collect();
                    writeln!(
                        f,
                        "{}Command (args: {:?})",
                        prefix, argv_texts
                    )?;
                }
                NdType::And => {
                    writeln!(f, "{}And", prefix)?;
                }
                NdType::Or => {
                    writeln!(f, "{}Or", prefix)?;
                }
                NdType::Redirection { .. } => {
                    writeln!(f, "{}Redir", prefix)?;
                }
                NdType::Pipe => {
                    writeln!(f, "{}Pipe", prefix)?;
                }
                NdType::PipeBoth => {
                    writeln!(f, "{}PipeBoth", prefix)?;
                }
                NdType::Cmdsep => {
                    writeln!(f, "{}Cmdsep", prefix)?;
                }
								NdType::NullNode => panic!()
            }
            Ok(())
        }

        write_tree(self, f, 0)
    }
}
