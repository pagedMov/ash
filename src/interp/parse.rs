use im::Vector;
use log::{error,debug,info,trace};

use crate::shellenv::ShellEnv;
use crate::interp::token::{tokenize, Tk, TkType};
use crate::interp::expand::expand;

#[derive(Debug,Clone)]
pub struct Conditional {
    pub condition: Box<Node>,
    pub body: Box<Node>
}

#[derive(Debug,Clone)]
pub struct CaseBlock {
    pub pattern: String,
    pub logic: Node
}

#[derive(Debug,Clone)]
pub enum Node {
    Root { deck: Vector<Node> },
    If { cond_blocks: Vector<Conditional>, else_block: Option<Box<Node>> },
    For { loop_vars: Vector<Tk>, loop_arr: Vector<Tk>, loop_body: Box<Node> },
    Loop { condition: bool, logic: Conditional },
    Case { input_var: Tk, cases: Vector<CaseBlock> },
    Select { select_var: Tk, opts: Vector<Tk>, body: Box<Node> },
    Pipeline { left: Box<Node>, right: Box<Node> },
    Chain { left: Box<Node>, right: Box<Node>, op: Box<Node> },
    BraceGroup { body: Box<Node> },
    Subshell { body: String }, // It's a string because we're going to parse it in a subshell later
    FuncDef { name: String, body: Box<Node> },
    Assignment {name: String, value: Option<String> },
    Command { argv: Vector<Tk>, redirs: Vector<Tk> },
    And,
    Or,
    Pipe,
    Cmdsep
}

#[derive(Debug, Clone)]
pub enum NdType {
    Root,
    If,
    For,
    Loop,
    Case,
    Select,
    Pipeline,
    Chain,
    Redir,
    BraceGroup,
    Subshell,
    FuncDef,
    Assignment,
    Command,
}

#[derive(Debug,Clone)]
pub struct ParseState<'a> {
    pub input: &'a str,
    pub shellenv: &'a ShellEnv,
    pub tokens: Vector<Tk>,
    pub ast: Node
}

pub fn descend<'a>(input: &'a str, shellenv: &'a ShellEnv) -> ParseState<'a> {
    info!("Starting descent into parsing with input: {:?}", input);
    let state = ParseState {
        input,
        shellenv,
        tokens: Vector::new(),
        ast: Node::Root { deck: Vector::new() },
    };

    let state = tokenize(state);

    let state = expand(state);

    let state = parse(state);

    state
}

pub fn parse(state: ParseState) -> ParseState {
    info!("Parsing tokens into AST...");
    let tokens = state.tokens.clone();
    debug!("Tokens before parsing: {:?}", tokens);

    let tokens = tokens.into_iter().collect::<Vec<Tk>>();
    trace!("Tokens collected into Vec: {:?}", tokens);

    let ast = get_tree(tokens.as_slice());
    debug!("Generated AST: {:#?}", ast);

    ParseState {
        input: state.input,
        shellenv: state.shellenv,
        tokens: state.tokens,
        ast
    }
}

pub fn get_tree(tokens: &[Tk]) -> Node {
    info!("Building AST from tokens...");
    let nodes = parse_linear(tokens);
    debug!("Linear nodes parsed: {:?}", nodes);

    let deck = join_at_operators(&nodes);
    debug!("AST deck after joining operators: {:?}", deck);

    Node::Root { deck }
}

pub fn parse_linear(tokens: &[Tk]) -> Vec<Node> {
    // First pass just makes nodes without joining at operators
    info!("Starting linear parsing of tokens...");
    let mut nodes = Vec::new();

    let mut tokens = tokens;
    while !tokens.is_empty() {
        info!("Current tokens: {:?}",tokens);
        use crate::interp::token::TkType::*;
        match tokens[0].class() {
            If => {
                info!("Found 'if' token, processing...");
                let (node, remaining_tokens) = build_if(tokens);
                nodes.push(node);
                tokens = remaining_tokens;
            }
            While => {
                info!("Found 'while' token, processing...");
                let (node, remaining_tokens) = build_loop(true, tokens);
                nodes.push(node);
                tokens = remaining_tokens;
            }
            Until => {
                info!("Found 'until' token, processing...");
                let (node, remaining_tokens) = build_loop(false, tokens);
                nodes.push(node);
                tokens = remaining_tokens;
            }
            For => {
                info!("Found 'for' token, processing...");
                let (node, remaining_tokens) = build_for(tokens);
                nodes.push(node);
                tokens = remaining_tokens;
            }
            Case => {
                info!("Found 'case' token, processing...");
                let (node, remaining_tokens) = build_case(tokens);
                nodes.push(node);
                tokens = remaining_tokens;
            }
            Select => {
                info!("Found 'select' token, processing...");
                let (node, remaining_tokens) = build_select(tokens);
                nodes.push(node);
                tokens = remaining_tokens;
            }
            Ident | String => {
                info!("Found command or string token, processing...");
                let (command_node, remaining_tokens) = build_command(tokens);
                nodes.push(command_node);
                tokens = remaining_tokens;
            }
            Cmdsep | LogicAnd | LogicOr | Pipe => {
                info!("Found operator token: {:?}, preserving as node", tokens[0].class());
                match tokens[0].class() {
                    Cmdsep => nodes.push(Node::Cmdsep),
                    LogicAnd => nodes.push(Node::And),
                    LogicOr => nodes.push(Node::Or),
                    Pipe => nodes.push(Node::Pipe),
                    _ => unreachable!()
                }
                tokens = &tokens[1..];
            }
            SOI => {
                trace!("Skipping Start of Input token");
                tokens = &tokens[1..];
            }
            EOI => {
                info!("End of Input token encountered, stopping parsing");
                break;
            }
            Elif | Do | Then | Else | Done | Fi => {
                tokens = &tokens[1..];
            }
            CaseSep | CaseDelim => {
                panic!("Found this outside of a case statement: {:?}",tokens[0]);
            }
            _ => {
                unimplemented!(
                    "Support for token type `{:?}` is not implemented yet",
                    tokens[0].class()
                );
            }
        }
        trace!("Current nodes: {:?}", nodes);
    }

    debug!("Completed linear parsing, nodes: {:?}", nodes);
    nodes
}

pub fn join_at_operators(deck: &[Node]) -> Vector<Node> {
    // Second pass will collect nodes at operators and construct pipelines and chains
    let mut buffer: Vector<Node> = Vector::new();
    let mut deck = Vector::from(deck);
    let mut left: Option<Node> = None;

    debug!("Starting join_at_operators with deck: {:?}", deck);

    while let Some(node) = deck.pop_front() {
        trace!("Processing node: {:?}", node);
        match node {
            Node::And | Node::Or => {
                debug!("Found logical operator: {:?}", node);
                if let Some(ref l_node) = left {
                    if let Some(r_node) = deck.pop_front() {
                        trace!("Creating Chain node with left: {:?}, operator: {:?}, right: {:?}", l_node, node, r_node);
                        let chain = Node::Chain {
                            left: Box::new(l_node.clone()),
                            right: Box::new(r_node),
                            op: Box::new(node),
                        };
                        left = Some(chain);
                    } else {
                        error!("Operator {:?} is missing a right operand", node);
                        panic!("Operator {:?} is missing a right operand", node);
                    }
                } else {
                    error!("Operator {:?} is missing a left operand", node);
                    panic!("Operator {:?} is missing a left operand", node);
                }
            }
            Node::Pipe => {
                debug!("Found Pipe operator");
                if let Some(ref l_node) = left {
                    if let Some(r_node) = deck.pop_front() {
                        trace!("Creating Pipeline node with left: {:?}, right: {:?}", l_node, r_node);
                        let pipeline = Node::Pipeline {
                            left: Box::new(l_node.clone()),
                            right: Box::new(r_node),
                        };
                        left = Some(pipeline);
                    } else {
                        error!("Pipe operator is missing a right operand");
                        panic!("Pipe operator is missing a right operand");
                    }
                } else {
                    error!("Pipe operator is missing a left operand");
                    panic!("Pipe operator is missing a left operand");
                }
            }
            Node::Cmdsep => {
                debug!("Found Cmdsep");
                if let Some(node) = left.take() {
                    trace!("Pushing node to buffer: {:?}", node);
                    buffer.push_back(node);
                }
            }
            _ => {
                debug!("Found general node: {:?}", node);
                if let Some(node) = left.take() {
                    trace!("Pushing node to buffer: {:?}", node);
                    buffer.push_back(node);
                }
                left = Some(node);
            }
        }
    }

    if let Some(node) = left {
        trace!("Pushing final node to buffer: {:?}", node);
        buffer.push_back(node);
    }

    debug!("Finished join_at_operators. Resulting buffer: {:?}", buffer);

    buffer
}

pub fn build_if(tokens: &[Tk]) -> (Node, &[Tk]) {
    let mut cond_blocks = Vector::new();
    let mut else_block = None;
    let mut index = 1; // Skip `if`

    while index < tokens.len() {
        // Extract condition tokens until `Then`
        let (cond_tokens, next_index) = extract_until(tokens, index, &[TkType::Then]);
        index = next_index + 1; // Skip `Then`

        // Extract body tokens until `Fi`, `Else`, or `Elif`
        let (body_tokens, next_index) = extract_until(tokens, index, &[TkType::Fi, TkType::Else, TkType::Elif]);
        index = next_index;

        // Add condition and body to `cond_blocks`
        let condition = Box::new(get_tree(cond_tokens));
        let body = Box::new(get_tree(body_tokens));
        cond_blocks.push_back(Conditional { condition, body });

        // Handle `Else` or terminate on `Fi`
        match tokens.get(index).map(|tk| tk.class()) {
            Some(TkType::Else) => {
                index += 1;
                let (body_tokens, next_index) = extract_until(tokens, index, &[TkType::Fi]);
                index = next_index + 1;
                else_block = Some(Box::new(get_tree(body_tokens)));
                break;
            }
            Some(TkType::Fi) => {
                index += 1;
                break;
            }
            _ => {} // Continue for `Elif`
        }
    }

    (
        Node::If { cond_blocks, else_block },
        &tokens[index..],
    )
}

fn extract_until<'a>(tokens: &'a [Tk], start: usize, stop_classes: &'a [TkType]) -> (&'a [Tk], usize) {
    let mut index = start;
    while index < tokens.len() && !stop_classes.contains(&tokens[index].class()) {
        index += 1;
    }
    (&tokens[start..index], index)
}
fn extract_until_text<'a>(tokens: &'a [Tk], start: usize, stop_pattern: &'a str) -> (&'a [Tk], usize) {
    let mut index = start;
    while index < tokens.len() && tokens[index].text() != stop_pattern {
        index += 1;
    }
    (&tokens[start..index], index)
}

pub fn build_for(tokens: &[Tk]) -> (Node,&[Tk]) {
    let mut index = 1;

    // TODO: figure out a way to get 'in' to actually be a keyword
    // Fix the logic in general so this code doesn't have to use awkward work arounds
    let (loop_vars,next_index) = extract_until_text(tokens, index, "in");
    index = next_index + 1;

    let (mut loop_arr,next_index) = extract_until(tokens, index, &[TkType::Do]);
    if loop_arr[loop_arr.len() - 1].text() == ";" {
        loop_arr = &loop_arr[..loop_arr.len() - 1]; // Cut the semicolon
    }
    index = next_index + 1;

    let (loop_body,next_index) = extract_until(tokens, index, &[TkType::Done]);
    index = next_index + 1;

    let loop_body = Box::new(get_tree(loop_body));

    (
        Node::For { loop_vars: loop_vars.into(), loop_arr: loop_arr.into(), loop_body },
        &tokens[index..]
    )
}

pub fn build_loop(condition: bool, tokens: &[Tk]) -> (Node, &[Tk]) {
    let loop_condition = condition;

    // Parse condition tokens until 'do'
    let mut index = 1; // Skip the loop keyword (e.g., 'while', 'until')
    let (cond_tokens,next_index) = extract_until(tokens, index, &[TkType::Do]);
    index = next_index + 1;

    let (body_tokens,next_index) = extract_until(tokens, index, &[TkType::Done]);
    index = next_index + 1;

    let condition = Box::new(get_tree(cond_tokens));
    let body = Box::new(get_tree(body_tokens));
    let logic = Conditional { condition, body };

    (
        Node::Loop { condition: loop_condition, logic },
        &tokens[index..]
    )
}

pub fn build_case(tokens: &[Tk]) -> (Node,&[Tk]) {
    // TODO: rewrite this to not use so much mutation
    let mut index = 1;
    let (input_var,next_index) = extract_until_text(tokens, index, "in");
    index = next_index + 1;
    if tokens[index].class() == TkType::Cmdsep {
        index += 1;
    }

    if input_var.len() != 1 {
        panic!("expected exactly one variable in case statement")
    }

    let (raw_cases,next_index) = extract_until(tokens,index,&[TkType::Esac]);
    let mut raw_cases = Vector::from(raw_cases);
    let mut cases = Vector::new();
    index = next_index + 1;
    if raw_cases[0].class() == TkType::Cmdsep {
        index += 1;
    }

    while !raw_cases.is_empty() {
        let pattern = raw_cases.pop_front().unwrap();
        if !matches!(pattern.class(),TkType::String | TkType::Ident) {
            panic!("Expected case pattern, got this: {:?} from {:?}",pattern.class(),raw_cases)
        }
        let pattern = pattern.text().to_string();
        if raw_cases.front().is_some_and(|tk| tk.class() != TkType::CaseSep) {
            panic!("Expected case separator, got this: {:?}",raw_cases.front().unwrap().class())
        } else {
            raw_cases.pop_front();
        }
        let mut block = Vec::new();
        while let Some(tk) = raw_cases.pop_front() {
            if tk.class() != TkType::CaseDelim {
                block.push(tk);
            } else { break }
        }
        let logic = get_tree(&block);
        let case = CaseBlock { pattern, logic };
        trace!("pushing case: {:?}",case);
        cases.push_back(case);
    }
    (Node::Case { input_var: input_var[0].clone(), cases }, &tokens[index..])
}

pub fn build_select(tokens: &[Tk]) -> (Node,&[Tk]) {
    let mut index = 1;

    // TODO: figure out a way to get 'in' to actually be a keyword
    // Fix the logic in general so this code doesn't have to use awkward work arounds
    let (select_var,next_index) = extract_until_text(tokens, index, "in");
    index = next_index + 1;
    if select_var.len() != 1 {
        panic!("expected exactly one variable in select statement")
    }
    let select_var = select_var[0].clone();

    let (mut opts,next_index) = extract_until(tokens, index, &[TkType::Do]);
    if opts[opts.len() - 1].class() == TkType::Cmdsep {
        opts = &opts[..opts.len() - 1]; // Cut the separator
    }
    index = next_index + 1;

    let (body,next_index) = extract_until(tokens, index, &[TkType::Done]);
    index = next_index + 1;

    let body = Box::new(get_tree(body));

    (
        Node::Select { select_var, opts: opts.into(), body },
        &tokens[index..]
    )

}

pub fn build_subshell(token: Tk) -> Node {
    Node::Subshell { body: token.text().into() }
}

pub fn build_func_def(tokens: &[Tk]) -> (Node,&[Tk]) {
    todo!("Implement build_func_def")
}

pub fn build_assignment(token: Tk) -> Node {
    let (var,val) = token.text().split_once('=').unwrap();
    Node::Assignment { name: var.to_string(), value: Some(val.to_string()) }
}

pub fn build_brace_group(tokens: &[Tk]) -> (Node,&[Tk]) {
    todo!("Implement build_brace_group")
}

pub fn build_pipeline<'a>(l_buffer: &'a [Tk], r_buffer: &'a [Tk]) -> (Node,&'a [Tk]) {
    todo!()
}

pub fn build_chain<'a>(l_buffer: &'a [Tk], r_buffer: &'a [Tk], op: TkType) -> (Node,&'a [Tk]) {
    todo!()
}

pub fn build_command(mut tokens: &[Tk]) -> (Node,&[Tk]) {
    let mut argv = Vector::new(); // Temporary mutable Vec to collect arguments
    let mut redirs = Vector::new(); // Temporary mutable Vec to collect redirections

    while !tokens.is_empty() {
        let tk = &tokens[0];

        match tk.class() {
            TkType::Cmdsep | TkType::LogicAnd | TkType::LogicOr | TkType::Pipe => {
                debug!("build_command breaking on: {:?}",tk);
                break
            }
            TkType::Ident | TkType::String => {
                // Add to argv
                argv.push_back(tk.clone());
            }
            TkType::Redirection { ref redir } => {
                // Handle redirection
                let mut new_redir = redir.clone();
                if redir.fd_target.is_none() {
                    if let Some(file_target) = tokens.first() {
                        match file_target.class() {
                            TkType::String | TkType::Ident => {
                                new_redir.file_target = Some(Box::new(file_target.clone()));
                                redirs.push_back(Tk {
                                    tk_type: TkType::Redirection { redir: new_redir },
                                    wd: tk.wd.clone(),
                                });
                                tokens = &tokens[1..]; // Advance the slice past file_target
                            }
                            _ => panic!("Expected identifier after redirection operator"),
                        }
                    } else {
                        panic!("Unexpected end of tokens after redirection operator");
                    }
                } else {
                    redirs.push_back(tk.clone());
                }
            }
            TkType::EOI | TkType::SOI => {},
            _ => {
                panic!("Unexpected token: {:?}", tk);
            }
        }
        tokens = &tokens[1..]; // Advance the slice
    }

    debug!("returning from build_command with tokens: {:?}",tokens);

    (Node::Command { argv, redirs, }, tokens)
}
