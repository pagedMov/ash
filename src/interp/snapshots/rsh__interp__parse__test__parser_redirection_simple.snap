---
source: src/interp/parse.rs
expression: flat_nodes
---
[
    Node {
        command: Some(
            Tk {
                tk_type: Ident,
                wd: WordDesc {
                    text: "echo",
                    span: Span {
                        start: 0,
                        end: 4,
                    },
                    flags: WdFlags(
                        BUILTIN,
                    ),
                },
            },
        ),
        nd_type: Builtin {
            argv: [
                Tk {
                    tk_type: Ident,
                    wd: WordDesc {
                        text: "echo",
                        span: Span {
                            start: 0,
                            end: 4,
                        },
                        flags: WdFlags(
                            BUILTIN,
                        ),
                    },
                },
                Tk {
                    tk_type: String,
                    wd: WordDesc {
                        text: "hello",
                        span: Span {
                            start: 5,
                            end: 10,
                        },
                        flags: WdFlags(
                            IS_ARG,
                        ),
                    },
                },
                Tk {
                    tk_type: String,
                    wd: WordDesc {
                        text: "world",
                        span: Span {
                            start: 11,
                            end: 16,
                        },
                        flags: WdFlags(
                            IS_ARG,
                        ),
                    },
                },
            ],
        },
        span: Span {
            start: 0,
            end: 16,
        },
        flags: NdFlags(
            VALID_OPERAND,
        ),
        redirs: [
            Node {
                command: None,
                nd_type: Redirection {
                    redir: Redir {
                        fd_source: 1,
                        op: Output,
                        fd_target: None,
                        file_target: Some(
                            Tk {
                                tk_type: String,
                                wd: WordDesc {
                                    text: "file.txt",
                                    span: Span {
                                        start: 19,
                                        end: 27,
                                    },
                                    flags: WdFlags(
                                        IS_ARG,
                                    ),
                                },
                            },
                        ),
                    },
                },
                span: Span {
                    start: 17,
                    end: 18,
                },
                flags: NdFlags(
                    IS_OP,
                ),
                redirs: [],
            },
        ],
    },
]
